pyccel/parser/semantic.py

```diff
@@ -670,47 +675,31 @@ def _infere_type(self, expr, **settings):
         d_var['cls_base'      ] = None
         d_var['cls_parameters'] = None
 
-
-
         # TODO improve => put settings as attribut of Parser
 
-        DEFAULT_FLOAT = settings.pop('default_float', 'real')
-
         if isinstance(expr, type(None)):
 
             return d_var
 
-        elif isinstance(expr, (Integer, int, PyccelArraySize)):
+        elif isinstance(expr, (Integer, Float, Complex, String,
+                               BooleanTrue, BooleanFalse,
+                               PyccelArraySize, Is, IndexedElement)):
 
-            d_var['datatype'   ] = 'int'
-            d_var['allocatable'] = False
-            d_var['rank'       ] = 0
-            d_var['precision'  ] = default_precision['int']
+            d_var['datatype'   ] = expr.dtype
+            d_var['allocatable'] = expr.rank>0
+            d_var['rank'       ] = expr.rank
+            d_var['precision'  ] = expr.precision
+            d_var['shape']       = expr.shape
             return d_var
+        elif expr in (PythonInt, PythonFloat, PythonComplex, PythonBool, NumpyInt, 
+                      Int32, Int64, NumpyComplex, Complex64, Complex128, NumpyFloat,
+                      Float64, Float32):
 
-        elif isinstance(expr, (Float, float)):
-
-            d_var['datatype'   ] = DEFAULT_FLOAT
-            d_var['allocatable'] = False
-            d_var['rank'       ] = 0
-            d_var['precision'  ] = default_precision['float']
-            return d_var
-
-        elif isinstance(expr, String):
-
-            d_var['datatype'   ] = 'str'
-            d_var['allocatable'] = False
-            d_var['rank'       ] = 0
-            return d_var
-
-        elif isinstance(expr, ImaginaryUnit):
-
-            d_var['datatype'   ] = 'complex'
-            d_var['allocatable'] = False
+            d_var['datatype'   ] = '*'
```

Comment:
What is a * datatype?

Reply:
it's a generic datatype, it's just a way of saying undetermined

Reply:
Why is the datatype undetermined here? Surely the datatype of a PythonInt is 'int' for example?

Reply:
it's not an 'int' it's a python data type so the real type will be type(int) and not int we can add that later but as long as it doesn't get generated we are fine.